[Safer CPP] Address remaining issues in several html/track classes

roberto-apple opened the pull request:
#### 6b5aa88ac433d9e407a2343d3b9d48b057f59930
<pre>
[Safer CPP] Address remaining issues in several html/track classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=302556">https://bugs.webkit.org/show_bug.cgi?id=302556</a>
<a href="https://rdar.apple.com/164758543">rdar://164758543</a>

Reviewed by Chris Dumez.

Address all remaining Safer CPP issues in several html/track classes.

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/html/track/AudioTrack.cpp:
(WebCore::AudioTrack::~AudioTrack):
(WebCore::AudioTrack::protectedPrivate const):
(WebCore::AudioTrack::setPrivate):
(WebCore::AudioTrack::setEnabled):
(WebCore::AudioTrack::inbandTrackIndex const):
(WebCore::AudioTrack::updateKindFromPrivate):
(WebCore::AudioTrack::setLogger):
* Source/WebCore/html/track/AudioTrack.h:
* Source/WebCore/html/track/DataCue.cpp:
(WebCore::DataCue::data const):
(WebCore::DataCue::cueContentsMatch const):
(WebCore::DataCue::value const):
* Source/WebCore/html/track/DataCue.h:
* Source/WebCore/html/track/InbandDataTextTrack.cpp:
(WebCore::InbandDataTextTrack::findIncompleteCue):
* Source/WebCore/html/track/InbandGenericTextTrack.cpp:
(WebCore::InbandGenericTextTrack::addGenericCue):
(WebCore::InbandGenericTextTrack::parser):
(WebCore::InbandGenericTextTrack::cueToExtend):
(WebCore::InbandGenericTextTrack::newRegionsParsed):
* Source/WebCore/html/track/InbandTextTrack.cpp:
(WebCore::InbandTextTrack::setPrivate):
(WebCore::InbandTextTrack::protectedPrivate const):
(WebCore::InbandTextTrack::setModeInternal):
(WebCore::InbandTextTrack::isClosedCaptions const):
(WebCore::InbandTextTrack::isSDH const):
(WebCore::InbandTextTrack::containsOnlyForcedSubtitles const):
(WebCore::InbandTextTrack::isMainProgramContent const):
(WebCore::InbandTextTrack::isEasyToRead const):
(WebCore::InbandTextTrack::isDefault const):
(WebCore::InbandTextTrack::inbandTrackIndex):
(WebCore::InbandTextTrack::inBandMetadataTrackDispatchType const):
(WebCore::InbandTextTrack::updateKindFromPrivate):
(WebCore::InbandTextTrack::startTimeVariance const):
(WebCore::InbandTextTrack::setLogger):
* Source/WebCore/html/track/InbandTextTrack.h:
* Source/WebCore/html/track/TextTrack.cpp:
(WebCore::TextTrack::protectedScriptExecutionContext const):
* Source/WebCore/html/track/TextTrack.h:
* Source/WebCore/html/track/TrackBase.h:
(WebCore::TrackBase::protectedLogger const):

Canonical link: <a href="https://commits.webkit.org/303176@main">https://commits.webkit.org/303176@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d986bf2565a2023595e4e70d5fad0be1ef459ee7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131512 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3844 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42477 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139019 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83291 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/77d1fb85-80be-4486-b16d-5407697b7ff7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133382 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3868 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3684 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100405 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f483ba44-40d7-4865-934c-25f9ba1c1608) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134458 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2783 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117718 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81205 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5d8b9c17-587c-4d07-a308-1b36dea292ff) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2696 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/435 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82211 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111315 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35843 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141665 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3588 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36361 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108771 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3636 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3193 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109004 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27631 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2717 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114048 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56776 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3649 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32453 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3474 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67060 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3730 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3579 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/track/AudioTrack.cpp:
```diff
@@ -79,26 +79,32 @@ AudioTrack::AudioTrack(ScriptExecutionContext* context, AudioTrackPrivate& track
+    removeClientFromTrackPrivateBase(protectedPrivate());
     m_private = trackPrivate;
-    m_private->setEnabled(m_enabled);
+    Ref privateRef = m_private;
```
just `private`, not `privateRef`. `privateRef` is not a common style in WebKit.

roberto-apple replied:
Agreed, but since `private` is reserved, that's why I chose `privateRef`
Would `audioTrackPrivate` suffice?

cdumez commented on Source/WebCore/html/track/DataCue.h:
```diff
@@ -57,6 +57,7 @@ class DataCue final : public TextTrackCue {
     void setData(JSC::ArrayBuffer&);
 
     const SerializedPlatformDataCue* platformValue() const { return m_platformValue.get(); }
+    const RefPtr<SerializedPlatformDataCue> protectedPlatformValue() const { return m_platformValue.get(); }
```
This should return a `RefPtr<const SerializedPlatformDataCue>`, not a `const RefPtr<SerializedPlatformDataCue>`.

cdumez commented on Source/WebCore/html/track/InbandGenericTextTrack.cpp:
```diff
@@ -213,7 +213,7 @@ RefPtr<TextTrackCue> InbandGenericTextTrack::cueToExtend(TextTrackCue& newCue)
     return [this, &newCue]() -> RefPtr<TextTrackCue> {
         for (size_t i = 0; i < m_cues->length(); ++i) {
-            RefPtr existingCue = m_cues->item(i);
+            RefPtr existingCue = RefPtr { m_cues }->item(i);
```
Please add a `Ref cues = m_cues;` before the loop and use `cues` in the loop condition and here. Calling `RefPtr { m_cues }` *inside* the loop is bad for refcounting churn.

cdumez commented on Source/WebCore/html/track/InbandGenericTextTrack.cpp:
```diff
@@ -262,7 +262,7 @@ void InbandGenericTextTrack::newCuesParsed()
 {
     for (auto& region : parser().takeRegions())
-        regions()->add(WTFMove(region));
+        RefPtr { regions() }->add(WTFMove(region));
```
Ditto. We want a local variable before the loop.

cdumez commented on Source/WebCore/html/track/TextTrack.h:
```diff
@@ -143,6 +143,7 @@ class TextTrack : public TrackBase, public EventTarget, public ActiveDOMObject {
     virtual void removeCuesNotInTimeRanges(const PlatformTimeRanges&);
 
     ScriptExecutionContext* scriptExecutionContext() const final;
+    RefPtr<ScriptExecutionContext> protectedScriptExecutionContext() const;
```
Why do we need this? This class already inherits a `protectedScriptExecutionContext()` function from ActiveDOMObject -> ContextDestructionObserver.

cdumez commented on Source/WebCore/html/track/TrackBase.h:
```diff
@@ -81,6 +81,7 @@ class TrackBase
 #if !RELEASE_LOG_DISABLED
     virtual void setLogger(const Logger&, uint64_t);
     const Logger& logger() const final { ASSERT(m_logger); return *m_logger.get(); }
+    RefPtr<const Logger> protectedLogger() const { return logger(); }
```
This should return a `Ref<const Logger>` since it cannot return null.

cdumez commented on Source/WebCore/html/track/InbandGenericTextTrack.cpp:
```diff
@@ -139,7 +139,7 @@ void InbandGenericTextTrack::addGenericCue(InbandGenericCue& inbandCue)
         return;
 
-    auto cue = TextTrackCueGeneric::create(*scriptExecutionContext(), inbandCue.startTime(), inbandCue.endTime(), inbandCue.content());
+    auto cue = TextTrackCueGeneric::create(*ActiveDOMObject::protectedScriptExecutionContext(), inbandCue.startTime(), inbandCue.endTime(), inbandCue.content());
```
Why do you need `ActiveDOMObject::` here? If we inherit `protectedScriptExecutionContext()` from 2 separate base classes, we should resolve the ambiguity using something like `using ActiveDOMObject::protectedScriptExecutionContext;` in the header.

roberto-apple replied:
Unless I'm missing something, dropping the `ActiveDOMObject::` qualifier and adding `using ActiveDOMObject::protectedScriptExecutionContext;` alone to the header (`InbandGenericTextTrack.h`) was leading to an ambiguous compilation error

```
../OpenSource/Source/WebCore/html/track/InbandGenericTextTrack.cpp:142:45: error: ambiguous conversion from derived class 'WebCore::InbandGenericTextTrack' to base class 'const WebCore::ContextDestructionObserver':
    class WebCore::InbandGenericTextTrack -> InbandTextTrack -> TextTrack -> TrackBase -> ContextDestructionObserver
  142 |     auto cue = TextTrackCueGeneric::create(*protectedScriptExecutionContext(), inbandCue.startTime(), inbandCue.endTime(), inbandCue.content());
  ```

In retrospect, since this was a single use, it probably would've just been easier from the beginning to wrap the normal scriptExecutionContext() in a RefPtr
  `auto cue = TextTrackCueGeneric::create(*RefPtr { scriptExecutionContext() }, inbandCue.startTime(), inbandCue.endTime(), inbandCue.content());`

cdumez commented on Source/WebCore/html/track/AudioTrack.cpp:
```diff
@@ -79,26 +79,32 @@ AudioTrack::AudioTrack(ScriptExecutionContext* context, AudioTrackPrivate& track
+    removeClientFromTrackPrivateBase(protectedPrivate());
     m_private = trackPrivate;
-    m_private->setEnabled(m_enabled);
+    Ref audioTrackPrivate = m_private;
```
I guess that's fine but then again, we don't need a local variable at all. I think we can just use the function parameter `trackPrivate` without protection.

cdumez commented on Source/WebCore/html/track/InbandGenericTextTrack.cpp:
```diff
@@ -261,8 +262,9 @@ void InbandGenericTextTrack::newCuesParsed()
+    RefPtr regions = this->regions();
     for (auto& region : parser().takeRegions())
-        regions()->add(WTFMove(region));
+        RefPtr { regions }->add(WTFMove(region));
```
Please use `regions` instead of `RefPtr { regions }`

cdumez reviewed (approved):
LGTM with fix.

roberto-apple committed:
[Safer CPP] Address remaining issues in several html/track classes
https://bugs.webkit.org/show_bug.cgi?id=302556
rdar://164758543

Reviewed by Chris Dumez.

Address all remaining Safer CPP issues in several html/track classes.

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/html/track/AudioTrack.cpp:
(WebCore::AudioTrack::~AudioTrack):
(WebCore::AudioTrack::protectedPrivate const):
(WebCore::AudioTrack::setPrivate):
(WebCore::AudioTrack::setEnabled):
(WebCore::AudioTrack::inbandTrackIndex const):
(WebCore::AudioTrack::updateKindFromPrivate):
(WebCore::AudioTrack::setLogger):
* Source/WebCore/html/track/AudioTrack.h:
* Source/WebCore/html/track/DataCue.cpp:
(WebCore::DataCue::data const):
(WebCore::DataCue::cueContentsMatch const):
(WebCore::DataCue::value const):
* Source/WebCore/html/track/DataCue.h:
* Source/WebCore/html/track/InbandDataTextTrack.cpp:
(WebCore::InbandDataTextTrack::findIncompleteCue):
* Source/WebCore/html/track/InbandGenericTextTrack.cpp:
(WebCore::InbandGenericTextTrack::addGenericCue):
(WebCore::InbandGenericTextTrack::parser):
(WebCore::InbandGenericTextTrack::cueToExtend):
(WebCore::InbandGenericTextTrack::newRegionsParsed):
* Source/WebCore/html/track/InbandTextTrack.cpp:
(WebCore::InbandTextTrack::setPrivate):
(WebCore::InbandTextTrack::protectedPrivate const):
(WebCore::InbandTextTrack::setModeInternal):
(WebCore::InbandTextTrack::isClosedCaptions const):
(WebCore::InbandTextTrack::isSDH const):
(WebCore::InbandTextTrack::containsOnlyForcedSubtitles const):
(WebCore::InbandTextTrack::isMainProgramContent const):
(WebCore::InbandTextTrack::isEasyToRead const):
(WebCore::InbandTextTrack::isDefault const):
(WebCore::InbandTextTrack::inbandTrackIndex):
(WebCore::InbandTextTrack::inBandMetadataTrackDispatchType const):
(WebCore::InbandTextTrack::updateKindFromPrivate):
(WebCore::InbandTextTrack::startTimeVariance const):
(WebCore::InbandTextTrack::setLogger):
* Source/WebCore/html/track/InbandTextTrack.h:
* Source/WebCore/html/track/TextTrack.cpp:
(WebCore::TextTrack::protectedScriptExecutionContext const):
* Source/WebCore/html/track/TextTrack.h:
* Source/WebCore/html/track/TrackBase.h:
(WebCore::TrackBase::protectedLogger const):

Canonical link: https://commits.webkit.org/303176@main
